File: ios/App/App/ScaleBuilder.swift
import Foundation

/// Construit les fréquences cibles d'une gamme musicale à partir d'une clé textuelle.
///
/// Entrée attendue : `"<note> <mode>"`, ex. `"F# minor"`, `"Bb major"`, `"C major"`.
/// Renvoie les fréquences en Hz (80–1200 Hz, octaves 2–5) triées par ordre croissant.
///
/// Toutes les méthodes sont pures (pas d'effet de bord) → testabilité maximale.
public enum ScaleBuilder {

    // ── Chromatique ────────────────────────────────────────────────────────────

    /// Index chromatique (0 = C … 11 = B) pour chaque notation courante.
    /// Inclut les enharmoniques et les altérations doubles les plus fréquentes.
    static let chromaticIndex: [String: Int] = [
        "C": 0,  "B#": 0,
        "C#": 1, "Db": 1,
        "D": 2,
        "D#": 3, "Eb": 3,
        "E": 4,  "Fb": 4,
        "F": 5,  "E#": 5,
        "F#": 6, "Gb": 6,
        "G": 7,
        "G#": 8, "Ab": 8,
        "A": 9,
        "A#": 10, "Bb": 10,
        "B": 11,  "Cb": 11,
    ]

    /// Intervalles en demi-tons de la gamme majeure (mode ionien).
    static let majorIntervals = [0, 2, 4, 5, 7, 9, 11]

    /// Intervalles en demi-tons de la gamme mineure naturelle (mode éolien).
    static let minorIntervals = [0, 2, 3, 5, 7, 8, 10]

    // ── API publique ───────────────────────────────────────────────────────────

    /// Fréquences cibles pour une clé donnée.
    /// - Parameter key: ex. `"F# minor"`, `"C major"`, `""` (→ tableau vide).
    /// - Returns: Fréquences triées en Hz dans [80, 1200].
    public static func frequencies(for key: String) -> [Float] {
        guard let (rootIndex, intervals) = parse(key: key) else { return [] }

        var result: [Float] = []
        result.reserveCapacity(intervals.count * 4)  // 4 octaves × N notes

        // Octaves 2 à 5 → C2 ≈ 65 Hz, B5 ≈ 988 Hz
        for octave in 2...5 {
            for interval in intervals {
                let midi = (octave + 1) * 12 + rootIndex + interval
                let hz   = 440.0 * powf(2.0, Float(midi - 69) / 12.0)
                if hz >= 80 && hz <= 1_200 { result.append(hz) }
            }
        }
        return result.sorted()
    }

    // ── Parsing ────────────────────────────────────────────────────────────────

    /// Parse la clé et retourne (rootIndex, intervals), ou nil si la clé est invalide.
    static func parse(key: String) -> (rootIndex: Int, intervals: [Int])? {
        let parts = key.trimmingCharacters(in: .whitespaces)
                       .components(separatedBy: .whitespaces)
                       .filter { !$0.isEmpty }

        guard let noteName = parts.first,
              let rootIndex = chromaticIndex[noteName] else { return nil }

        let mode = parts.dropFirst().joined(separator: " ").lowercased()
        let intervals = mode.contains("min") ? minorIntervals : majorIntervals

        return (rootIndex, intervals)
    }
}

File: android/app/src/main/java/net/laprod/app/ScaleBuilder.kt
package net.laprod.app

import kotlin.math.abs
import kotlin.math.log2
import kotlin.math.pow

/**
 * Construit les fréquences cibles d'une gamme musicale.
 *
 * Entrée : `"<note> <mode>"` ex. `"F# minor"`, `"Bb major"`, `"C major"`.
 * Sortie  : tableau de fréquences en Hz (80–1200 Hz, octaves 2–5), trié par ordre croissant.
 *
 * Toutes les méthodes sont pures → testabilité maximale sans Android SDK.
 */
object ScaleBuilder {

    // ── Chromatique ────────────────────────────────────────────────────────────

    private val CHROMATIC_INDEX: Map<String, Int> = mapOf(
        "C"  to  0, "B#" to  0,
        "C#" to  1, "Db" to  1,
        "D"  to  2,
        "D#" to  3, "Eb" to  3,
        "E"  to  4, "Fb" to  4,
        "F"  to  5, "E#" to  5,
        "F#" to  6, "Gb" to  6,
        "G"  to  7,
        "G#" to  8, "Ab" to  8,
        "A"  to  9,
        "A#" to 10, "Bb" to 10,
        "B"  to 11, "Cb" to 11,
    )

    private val MAJOR_INTERVALS = intArrayOf(0, 2, 4, 5, 7, 9, 11)
    private val MINOR_INTERVALS = intArrayOf(0, 2, 3, 5, 7, 8, 10)

    private const val FREQ_MIN = 80f
    private const val FREQ_MAX = 1_200f

    // ── API publique ───────────────────────────────────────────────────────────

    /**
     * Retourne les fréquences cibles pour une clé donnée.
     * @param key ex. "F# minor", "C major", "" → tableau vide.
     */
    fun buildScaleHz(key: String): FloatArray {
        val (rootIndex, intervals) = parse(key) ?: return FloatArray(0)

        val result = mutableListOf<Float>()
        for (octave in 2..5) {
            for (interval in intervals) {
                val midi = (octave + 1) * 12 + rootIndex + interval
                val hz   = (440.0 * 2.0.pow((midi - 69).toDouble() / 12.0)).toFloat()
                if (hz in FREQ_MIN..FREQ_MAX) result.add(hz)
            }
        }
        return result.toFloatArray().also { it.sort() }
    }

    // ── Parsing ────────────────────────────────────────────────────────────────

    /**
     * Parse la clé et retourne (rootIndex, intervalles), ou null si invalide.
     * Exposé interne pour les tests.
     */
    internal fun parse(key: String): Pair<Int, IntArray>? {
        val parts = key.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
        val noteName  = parts.firstOrNull() ?: return null
        val rootIndex = CHROMATIC_INDEX[noteName]  ?: return null
        val mode      = parts.drop(1).joinToString(" ").lowercase()
        val intervals = if ("min" in mode) MINOR_INTERVALS else MAJOR_INTERVALS
        return Pair(rootIndex, intervals)
    }
}
